tool/clean/clean_host/clean_host.go

```diff
@@ -0,0 +1,98 @@
+
+
+package main
+
+import (
+	"context"
+	"github.com/aws/amazon-cloudwatch-agent/tool/clean"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go/aws"
+	"log"
+	"time"
+)
+
+// Clean integration hosts if they have been open longer than 1 day
+func main() {
+	err := cleanHost()
+	if err != nil {
+		log.Fatalf("errors cleaning %v", err)
+	}
+}
+
+func cleanHost() error {
+	log.Print("Begin to clean EC2 Host")
+
+	cxt := context.Background()
+	defaultConfig, err := config.LoadDefaultConfig(cxt)
+	if err != nil {
+		return err
+	}
+	ec2client := ec2.NewFromConfig(defaultConfig)
+
+	terminateInstances(cxt, ec2client)
+	return err
+}
+
+func terminateInstances(cxt context.Context, ec2client *ec2.Client) {
+	maxResults := int32(1000)
+	const tagName = "tag:Name"
+	var tagValue = []string{
+		"buildLinuxPackage",
+		"buildPKG",
+		"buildMSI",
+		"MSIUpgrade_*",
+		"IntegrationTestBase",
+		"CWADockerImageBuilderX86",
+		"CWADockerImageBuilderARM64",
+	}
+	nameFilter := types.Filter{Name: aws.String(tagName), Values: tagValue}
+
+	nextToken := aws.String("")
+	instanceInput := ec2.DescribeInstancesInput{
+		Filters: []types.Filter{
+			nameFilter,
+			{Name: aws.String("instance-state-name"),
+				Values: []string{"running"}}},
+		MaxResults: maxResults}
+	for {
+		instanceIds := make([]string, 0)
+		expirationDateInstance := time.Now().UTC().Add(clean.KeepDurationOneDay)
+		describeInstanceOutput, _ := ec2client.DescribeInstances(cxt, &instanceInput)
+		for _, reservation := range describeInstanceOutput.Reservations {
+			for _, instance := range reservation.Instances {
+				log.Printf("instance id %v experation date %v host creation date raw %v host state %v",
+					*instance.InstanceId, expirationDateInstance, *instance.LaunchTime, instance.State)
+				if expirationDateInstance.After(*instance.LaunchTime) {
+					log.Printf("Try to delete instance %s tags %v launch-date %s", *instance.InstanceId, instance.Tags, *instance.LaunchTime)
+					instanceIds = append(instanceIds, *instance.InstanceId)
+				}
+			}
+		}
+		if len(instanceIds) == 0 {
+			log.Printf("No instances to terminate")
+			return
+		}
+
+		log.Printf("instances to temrinate %v", instanceIds)
+		terminateInstance := ec2.TerminateInstancesInput{InstanceIds: instanceIds}
+		_, err := ec2client.TerminateInstances(cxt, &terminateInstance)
+		if err != nil {
+			log.Printf("Error %v terminating instances %v", err, instanceIds)
+		}
+		nextToken = describeInstanceOutput.NextToken
+		if describeInstanceOutput.NextToken == nil {
+			break
+		}
+		instanceInput = ec2.DescribeInstancesInput{Filters: []types.Filter{
+				nameFilter,
+				{Name: aws.String("instance-state-name"), Values: []string{"running"}}},
+			MaxResults: maxResults,
+			NextToken: nextToken}
+		// prevent throttle https://docs.aws.amazon.com/AWSEC2/latest/APIReference/throttling.html
+		time.Sleep(time.Minute)
```

Comment:
Throttle is within second, so its better to reduce the time here (e.g 10 * time.Second perhaps)

Reply:
The docs for 1000 are 1 minute based. Read the doc.

Reply:
```
100 Describe* requests in one second. so you can make up to 100 Describe* requests in one second. If you exceed 100 requests in a second, you are throttled and the remaining requests within that second fail.
```

Reply:
Since its a non-mutating actions, its fallin within this category - only has 100 Bucket Maximum Capacity.

Reply:
```
For example, the resource token bucket size for RunInstances is 1000 tokens, and the refill rate is two tokens per second. Therefore, you can immediately launch 1000 instances, using any number of API requests, such as one request for 1000 instances or four requests for 250 instances. After the resource token bucket is empty, you can launch up to two instances every second, using either one request for two instances or two requests for one instance.
```

Reply:
Its rare use case for me IMO and maybe I'm focusing too much on `request rate` not `resources rate`. Deleting 1000 EC2 instances for the first call and 200 instances for every next call would be fine for me ( in term of cost, its small to neglect it). However, its not worth to block it.